indexing-service/src/main/java/io/druid/indexing/common/IndexTaskClient.java

```diff
@@ -0,0 +1,381 @@
+              port,
+              path,
+              query,
+              null
+          );
+          request = new Request(method, serviceUri.toURL());
+
+          // used to validate that we are talking to the correct worker
+          request.addHeader(ChatHandlerResource.TASK_ID_HEADER, taskId);
+
+          if (content.length > 0) {
+            request.setContent(Preconditions.checkNotNull(mediaType, "mediaType"), content);
+          }
+
+          log.debug("HTTP %s: %s", method.getName(), serviceUri.toString());
+          response = httpClient.go(request, new FullResponseHandler(StandardCharsets.UTF_8), httpTimeout).get();
+        }
+        catch (IOException | ChannelException ioce) {
+          throw ioce;
+        }
+        catch (InterruptedException ie) {
+          Thread.currentThread().interrupt();
+          throw new RuntimeException(ie);
+        }
+        catch (Exception e) {
+          throw new RuntimeException(e);
+        }
+
+        int responseCode = response.getStatus().getCode();
+        if (responseCode / 100 == 2) {
+          return response;
+        } else if (responseCode == 400) { // don't bother retrying if it's a bad request
+          throw new IAE("Received 400 Bad Request with body: %s", response.getContent());
+        } else {
+          throw new IOE("Received status [%d]", responseCode);
+        }
+      }
+      catch (IOException | ChannelException e) {
+
+        // Since workers are free to move tasks around to different ports, there is a chance that a task may have been
+        // moved but our view of its location has not been updated yet from ZK. To detect this case, we send a header
+        // identifying our expected recipient in the request; if this doesn't correspond to the worker we messaged, the
+        // worker will return an HTTP 404 with its ID in the response header. If we get a mismatching task ID, then
+        // we will wait for a short period then retry the request indefinitely, expecting the task's location to
+        // eventually be updated.
+
+        final Duration delay;
+        if (response != null && response.getStatus().equals(HttpResponseStatus.NOT_FOUND)) {
+          String headerId = response.getResponse().headers().get(ChatHandlerResource.TASK_ID_HEADER);
+          if (headerId != null && !headerId.equals(taskId)) {
+            log.warn(
+                "Expected worker to have taskId [%s] but has taskId [%s], will retry in [%d]s",
+                taskId, headerId, TASK_MISMATCH_RETRY_DELAY_SECONDS
+            );
+            delay = Duration.standardSeconds(TASK_MISMATCH_RETRY_DELAY_SECONDS);
+          } else {
+            delay = retryPolicy.getAndIncrementRetryDelay();
+          }
+        } else {
+          delay = retryPolicy.getAndIncrementRetryDelay();
+        }
+        String urlForLog = (request != null
+                            ? request.getUrl().toString()
+                            : StringUtils.format(
```

Comment:
This was originally in KafkaIndexTaskClient and moved to here. I guess this code isn't actually creating an URL, but creating a String instance for logging instead.

Reply:
ok, in that case it would make sense to never fail if possible. A potential change that could make sense is to use the non strict format method in StringUtils, but is an optional change.

Reply:
That sounds good. Changed.